File: VB/TutorialsCustomItems/Form1.vb
Imports DevExpress.DashboardCommon
Imports DevExpress.DashboardCommon.ViewerData
Imports DevExpress.Spreadsheet
Imports TutorialsCustomItems.CustomItems

Namespace TutorialsCustomItems
	Partial Public Class Form1
		Inherits Form

		Public Sub New()
			InitializeComponent()
			dashboardDesigner1.CreateRibbon()
			dashboardDesigner1.LoadDashboard("..\..\Data\TutorialCustomItems.xml")
			dashboardDesigner1.CreateCustomItemBars()
			AddHandler dashboardDesigner1.CustomizeExportDocument, AddressOf dashboardDesigner1_CustomizeExportDocument
		End Sub
		Private Sub DashboardDesigner1_DataLoading(ByVal sender As Object, ByVal e As DevExpress.DashboardCommon.DataLoadingEventArgs) Handles dashboardDesigner1.DataLoading
			If e.DataSourceComponentName = "dsExport" Then
				e.Data = CustomItemDataGenerator.GetExportData()
			ElseIf e.DataSourceComponentName = "dsContinent" Then
				e.Data = CustomItemDataGenerator.GetContinentData()
			End If
		End Sub

		Private Sub DashboardDesigner1_CustomDashboardItemControlCreating(ByVal sender As Object, ByVal e As DevExpress.DashboardWin.CustomDashboardItemControlCreatingEventArgs) Handles dashboardDesigner1.CustomDashboardItemControlCreating
			If e.MetadataType Is GetType(HelloWorldItemMetadata) Then
				e.CustomControlProvider = New HelloWorldItemProvider()
			End If
			If e.MetadataType Is GetType(SimpleTableMetadata) Then
				e.CustomControlProvider = New SimpleTableProvider()
			End If
			If e.MetadataType Is GetType(FunnelItemMetadata) Then
				e.CustomControlProvider = New FunnelItemControlProvider(TryCast(dashboardDesigner1.Dashboard.Items(e.DashboardItemName), CustomDashboardItem(Of FunnelItemMetadata)))
			End If
		End Sub

		Private Sub DashboardDesigner1_Load(ByVal sender As Object, ByVal e As EventArgs) Handles dashboardDesigner1.Load
		End Sub

		Private Sub dashboardDesigner1_CustomizeExportDocument(ByVal sender As Object, ByVal e As CustomizeExportDocumentEventArgs)
			Dim item As CustomDashboardItem = TryCast(dashboardDesigner1.Dashboard.Items.FirstOrDefault(Function(i) i.ComponentName = e.ItemComponentName), CustomDashboardItem)

			If item IsNot Nothing Then
				Dim workbook As New DevExpress.Spreadsheet.Workbook()
				Dim worksheet As Worksheet = workbook.Worksheets(0)

				Dim itemData As MultiDimensionalData = e.GetItemData(e.ItemComponentName)
				Dim customItemData As New CustomItemData(item, itemData)

				Dim flatData As DashboardFlatDataSource = customItemData.GetFlatData()
				Dim columns As IList(Of DashboardFlatDataColumn) = flatData.GetColumns()
				For colIndex As Integer = 0 To columns.Count - 1
					worksheet.Cells(0, colIndex).Value = columns(colIndex).DisplayName
					worksheet.Cells(0, colIndex).FillColor = Color.LightGreen
					worksheet.Cells(0, colIndex).Font.FontStyle = SpreadsheetFontStyle.Bold
					Dim headerOffset As Integer = 1
					For rowIndex As Integer = 0 To flatData.Count - 1
						worksheet.Cells(rowIndex + headerOffset, colIndex).Value = flatData.GetDisplayText(columns(colIndex).Name, rowIndex)
					Next rowIndex
				Next colIndex
				e.Stream.SetLength(0)
				workbook.SaveDocument(e.Stream, DocumentFormat.Xlsx)
			End If
		End Sub
	End Class
End Namespace

File: CS/TutorialsCustomItems/Form1.cs
using System;
using System.Collections.Generic;
using System.Drawing;
using System.Linq;
using System.Windows.Forms;
using DevExpress.DashboardCommon;
using DevExpress.DashboardCommon.ViewerData;
using TutorialsCustomItems.CustomItems;
using DevExpress.Spreadsheet;

namespace TutorialsCustomItems {
    public partial class Form1 : Form {
        public Form1() {
            InitializeComponent();
            dashboardDesigner1.CreateRibbon();
            dashboardDesigner1.LoadDashboard(@"..\..\Data\TutorialCustomItems.xml");
            dashboardDesigner1.CreateCustomItemBars();
            dashboardDesigner1.CustomizeExportDocument += dashboardDesigner1_CustomizeExportDocument;
        }

        private void DashboardDesigner1_DataLoading(object sender, DevExpress.DashboardCommon.DataLoadingEventArgs e){
            if (e.DataSourceComponentName == "dsExport")
                e.Data = CustomItemDataGenerator.GetExportData();
            else if (e.DataSourceComponentName == "dsContinent")
                e.Data = CustomItemDataGenerator.GetContinentData();
        }

        private void DashboardDesigner1_CustomDashboardItemControlCreating(object sender, DevExpress.DashboardWin.CustomDashboardItemControlCreatingEventArgs e) {
            if (e.MetadataType == typeof(HelloWorldItemMetadata))
                e.CustomControlProvider = new HelloWorldItemProvider();
            if (e.MetadataType == typeof(SimpleTableMetadata))
                e.CustomControlProvider = new SimpleTableProvider();
            if (e.MetadataType == typeof(FunnelItemMetadata))
                e.CustomControlProvider = new FunnelItemControlProvider(dashboardDesigner1.Dashboard.Items[e.DashboardItemName] as CustomDashboardItem<FunnelItemMetadata>);
        }

        private void dashboardDesigner1_CustomizeExportDocument(object sender, CustomizeExportDocumentEventArgs e) {
            CustomDashboardItem item = dashboardDesigner1.Dashboard.Items.FirstOrDefault(i => i.ComponentName == e.ItemComponentName) as CustomDashboardItem;

            if (item != null) {
                DevExpress.Spreadsheet.Workbook workbook = new DevExpress.Spreadsheet.Workbook();
                Worksheet worksheet = workbook.Worksheets[0];

                MultiDimensionalData itemData = e.GetItemData(e.ItemComponentName);
                CustomItemData customItemData = new CustomItemData(item, itemData);

                DashboardFlatDataSource flatData = customItemData.GetFlatData();
                IList<DashboardFlatDataColumn> columns = flatData.GetColumns();
                for (int colIndex = 0; colIndex < columns.Count; colIndex++) {
                    worksheet.Cells[0, colIndex].Value = columns[colIndex].DisplayName;
                    worksheet.Cells[0, colIndex].FillColor = Color.LightGreen;
                    worksheet.Cells[0, colIndex].Font.FontStyle = SpreadsheetFontStyle.Bold;
                    int headerOffset = 1;
                    for (int rowIndex = 0; rowIndex < flatData.Count; rowIndex++)
                        worksheet.Cells[rowIndex + headerOffset, colIndex].Value = flatData.GetDisplayText(columns[colIndex].Name, rowIndex);
                }
                e.Stream.SetLength(0);
                workbook.SaveDocument(e.Stream, DocumentFormat.Xlsx);
            }
        }
    }
}
